src/core/net/dns_client.cpp

```diff
@@ -199,6 +198,23 @@ void Client::SendCopy(const Message &aMessage, const Ip6::MessageInfo &aMessageI
     }
 }
 
+uint16_t Client::GenerateUniqueRandomId(void)
+{
+    uint16_t randomId;
+    otError  error;
+
+    OT_UNUSED_VARIABLE(error);
+
+    do
+    {
+        error = Random::Crypto::FillBuffer(reinterpret_cast<uint8_t *>(&randomId), sizeof(randomId));
+        OT_ASSERT(error == OT_ERROR_NONE);
```

Comment:
Is generating a DNS transaction id so critical that we should assert here? Crypto sources (library or other core) may need to gather more entropy or be blocked by another request. This could lead to random failures. Maybe retry mechanism (could be added at crypto layer) or reporting an error during and dropping the query would be better?

Reply:
Current DNS Client implementation facilitates DNS cache poisoning attacks by implementing deterministic way of assigning IDs (just increment ID by one).

One option to remove assert would be to use PRNG, however, the attacker may send a series of fake replies with several ID numbers trying to guess the proper one. If ID is generated with poor Pseudo-Random Number Generator, attacker has higher probability of success. While i personally didn't do any calculation with use of our PRNG, there are few papers in the internet suggesting to use TRNG mechanism.

`Crypto` RNG module is used without error checking in few places in OpenThread, in other places assertion like here is implemented. I would say, that we can improve it in subsequent PR, for example by implementing a backup mechanism to use `NonCrypto` module in case `Crypto` fails. I believe simple retry mechanism for `Crypto` may not work always, and in some situation may lead to infinite loop. If that does not work, I can re-write the function to return error code. @abtink any suggestion?

Reply:
> implementing a backup mechanism to use NonCrypto module in case Crypto fails
We should not substitute the results of `Crypto` functions with `NonCrypto` ones since this may lead to security vulnerabilities. 

Additionally, OT_ASSERT may be disabled. In such a case the current implementation of this particular function (and maybe in some other places as you pointed out) will just ignore an error and return some bytes but not necessarily random.

Reply:
It looks like a topic worth discussing in a separate ticket.

Reply:
@konradderda in my last commit i removed assert and return error code now instead. Please have a look.